bridges/otelslog/handler.go

```diff
@@ -140,13 +218,258 @@ func (h *Handler) Enabled(ctx context.Context, l slog.Level) bool {
+	//     Value: []KeyValue{{
+	//       Key: "H",
+	//       Value: []KeyValue{{
+	//         Key: "I",
+	//         Value: []KeyValue{
+	//           String("key", "value"),
+	//         },
+	//       }},
+	//     }},
+	//   }
+	//
+	// Therefore, groups are structured as a linked-list with the "leaf" node
+	// being the head of the list. Following the above example, the group data
+	// representation would be ...
+	//
+	//   *group{"I", next: *group{"H", next: *group{"G"}}}
+	next *group
+}
+
+// NextNonEmpty returns the next group within g's linked-list that has
+// attributes (including g itself). If no group is found, nil is returned.
+func (g *group) NextNonEmpty() *group {
+	if g == nil || g.attrs.Len() > 0 {
+		return g
+	}
+	return g.next.NextNonEmpty()
+}
+
+// KeyValue returns group g containing kvs as a [log.KeyValue]. The value of
+// the returned KeyValue will be of type [log.KindMap].
+//
+// The passed kvs are rendered in the returned value, but are not added to the
+// group.
+//
+// This does not check g. It is the callers responsibility to ensure g is
+// non-empty or kvs is non-empty so as to return a valid group representation
+// (according to slog).
+func (g *group) KeyValue(kvs ...log.KeyValue) log.KeyValue {
+	// Assumes checking of group g already performed (i.e. non-empty).
+	out := log.Map(g.name, g.attrs.KeyValues(kvs...)...)
+	g = g.next
+	for g != nil {
+		// A Handler should not output groups if there are no attributes.
+		if g.attrs.Len() > 0 {
+			out = log.Map(g.name, g.attrs.KeyValues(out)...)
+		}
+		g = g.next
+	}
+	return out
+}
+
+// Clone returns a copy of g.
+func (g *group) Clone() *group {
+	if g == nil {
+		return g
+	}
+	g2 := *g
+	g2.attrs = g2.attrs.Clone()
+	return &g2
+}
+
+// AddAttrs add attrs to g.
+func (g *group) AddAttrs(attrs []slog.Attr) {
+	if g.attrs == nil {
+		g.attrs = newKVBuffer(len(attrs))
+	}
+	g.attrs.AddAttrs(attrs)
+}
+
+var kvBufferPool = sync.Pool{
+	New: func() any {
+		// Based on slog research (https://go.dev/blog/slog#performance), 95%
+		// of use-cases will use 5 or less attributes.
+		return newKVBuffer(5)
+	},
+}
+
+func getKVBuffer() (buf *kvBuffer, free func()) {
+	buf = kvBufferPool.Get().(*kvBuffer)
+	return buf, func() {
+		// TODO: limit returned size so the pool doesn't hold on to very large
+		// buffers. Idea is based on
+		// https://cs.opensource.google/go/x/exp/+/814bf88c:slog/internal/buffer/buffer.go;l=27-34
+
+		// Do not modify any previously held data.
+		buf.data = buf.data[:0:0]
+		kvBufferPool.Put(buf)
+	}
+}
+
+type kvBuffer struct {
+	data []log.KeyValue
+}
+
+func newKVBuffer(n int) *kvBuffer {
+	return &kvBuffer{data: make([]log.KeyValue, 0, n)}
+}
+
+// Len returns the number of [log.KeyValue] held by b.
+func (b *kvBuffer) Len() int {
+	if b == nil {
+		return 0
+	}
+	return len(b.data)
+}
+
+// Clone returns a copy of b.
+func (b *kvBuffer) Clone() *kvBuffer {
+	if b == nil {
+		return nil
+	}
+	return &kvBuffer{data: slices.Clone(b.data)}
```

Comment:
You lose any extra capacity when you clone b.data, so the subsequent AddAttrs will always allocate. (I think.)

In general though, it's probably best not to worry too much about these details. You can always tweak the performance later.